sdk/core/core/src/az_span.c

```diff
@@ -153,6 +154,30 @@ AZ_NODISCARD az_result az_span_atou32(az_span span, uint32_t* out_number)
   return AZ_OK;
 }
 
+AZ_NODISCARD az_result az_span_atod(az_span span, az_span* conversion_buffer, double* out_number)
+{
+  AZ_PRECONDITION_VALID_SPAN(span, 1, false);
+  AZ_PRECONDITION_NOT_NULL(out_number);
+
+  if (conversion_buffer == NULL)
+  {
+    // This means there is already an invalid character after the last digit to convert
+    *out_number = atof((char*)az_span_ptr(span));
+    return AZ_OK;
+  }
+
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(
+      *conversion_buffer, az_span_size(span) + 1 /*plus 1 for the extra char to be added*/);
+
+  // Copy content to conversion buffer
+  az_span remainder = az_span_copy(*conversion_buffer, span);
```

Comment:
What happens in this case? And what should happen?
```C
uint8 buffer[2147483647] = { '1', '2', '3' }; // The number 123
az_span source = AZ_SPAN_FROM_BUFFER(buffer);

uint8 conversion[2];
az_span conv_span = AZ_SPAN_FROM_BUFFER(conversion);

double d;
az_result res = az_span_atod(source, &conv_span, &d);
```

Reply:
AZ_ERROR_INSUFFICIENT_SPAN_SIZE, there is a tests about it and it's documented, it means the conversion buffer is not big enough

Reply:
Did you try it (especially with preconditions off)? That is not what actually happens given the current implementation.

We hit a precondition failure `az_span_copy`
OR
When preconditions are off, we return AZ_OK with 12.0 as the double value.
```C

static void test_az_span_atod_fail(void** state)
{
  (void)state;

  uint8_t* p_array = (uint8_t*)malloc(2147483647);
  p_array[0] = '1';
  p_array[1] = '2';
  p_array[2] = '3'; // The number 123

  az_span source = az_span_init(p_array, 2147483647);

  uint8_t conversion[2];
  az_span conv_span = AZ_SPAN_FROM_BUFFER(conversion);

  double d;
  az_result res = az_span_atod(source, &conv_span, &d);
  printf("Double: %lf\n", d);
  assert_true(res == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);

  free(p_array);
}
```